packages/lesmis/src/views/MapView.js

```diff
@@ -17,27 +33,137 @@ const Container = styled.div`
   const { entityCode } = useUrlParams();
+  const [activeTileSetKey, setActiveTileSetKey] = useState(TILE_SETS[0].key);
+  const [hiddenMeasures, setHiddenMeasures] = useState([]);
   const [selectedOverlay, setSelectedOverlay] = useState(null);
 
+  useEffect(() => {
+    setHiddenMeasures([]);
+  }, [setHiddenMeasures, selectedOverlay]);
+
+  const { data: entityData } = useEntityData({
+    entityCode,
+  });
+
+  const { data: overlayReportData } = useMapOverlayReportData({
+    entityCode,
+    mapOverlayCode: selectedOverlay,
+    hiddenMeasures,
+  });
+
   const { data: overlaysData, isLoading } = useMapOverlaysData({ entityCode });
 
+  const { data: provicesData } = useProvinceData();
+
+  const handleLocationChange = (map, bounds) => {
+    const mapBoxBounds = L.latLngBounds(bounds);
+    const maxBounds = mapBoxBounds.pad(2);
+    map.setMaxBounds(maxBounds);
+  };
+
+  const handleSetHiddenMeasures = value => {
+    const exists = hiddenMeasures.includes(value);
+
+    if (exists) {
+      const newHiddenMeasures = hiddenMeasures.filter(m => m !== value);
+      setHiddenMeasures(newHiddenMeasures);
+    } else {
+      const newHiddenMeasures = [...hiddenMeasures, value];
+      setHiddenMeasures(newHiddenMeasures);
+    }
+  };
+
+  const activeTileSet = TILE_SETS.find(tileSet => tileSet.key === activeTileSetKey);
+  const isCountry = entityData?.type === 'country';
+
   return (
     <Container>
-      <MapOverlaysPanel
-        isLoading={isLoading}
-        overlays={overlaysData}
-        selectedOverlay={selectedOverlay}
-        setSelectedOverlay={setSelectedOverlay}
-      />
-      <Main>Map</Main>
+      {isCountry ? (
+        <EmptyPanel>Select a province from the map to view data.</EmptyPanel>
+      ) : (
+        <MapOverlaysPanel
+          isLoading={isLoading}
+          overlays={overlaysData}
+          selectedOverlay={selectedOverlay}
+          setSelectedOverlay={setSelectedOverlay}
+        />
+      )}
+      <Main>
+        <Map
+          bounds={entityData ? entityData.bounds : null}
+          onLocationChange={handleLocationChange}
+          dragging
+        >
+          <TileLayer tileSetUrl={activeTileSet.url} />
+          {isCountry ? (
+            <PolygonLayer entities={provicesData} Polygon={EntityPolygonLink} />
+          ) : (
+            <>
+              <MarkerLayer
+                measureData={overlayReportData ? overlayReportData.measureData : null}
+                measureOptions={overlayReportData ? overlayReportData.measureOptions : null}
+              />
+              <Polygon entity={entityData} />
+            </>
+          )}
+        </Map>
+        <TilePicker
+          tileSets={TILE_SETS}
+          activeTileSet={activeTileSet}
+          onChange={setActiveTileSetKey}
+        />
+        {!isCountry && (
+          <Legend
+            measureOptions={overlayReportData?.measureOptions}
+            setHiddenMeasures={handleSetHiddenMeasures}
```

Comment:
I've stopped insisting that we should use `useCallback` absolutely everywhere, but I think this function gets passed down through a long enough chain of components that forcing them to re-render each render cycle of this parent component isn't worth it. Let's convert `handleSetHiddenMeasures` to a `useCallback` plz